From 6ed646d1c21e0ef54a6966a8f54f8bb7dbed2b99 Mon Sep 17 00:00:00 2001
From: "Thulasimani,Sivakumar" <sivakumar.thulasimani@intel.com>
Date: Mon, 11 Jan 2016 15:04:59 +0530
Subject: [PATCH 1/5] FOR_UPSTREAM [VPG]: drm/i915: Add DCOAMP support for CHT

New registers were added to improve the stability of
clock generation.
This patch adds DCOAMP programming
to appropriate DPIO register as part of DPIO clock
programming.

Change-Id: I30dbce29bf20a9ab4314bb2b35a96fc82cc408d2
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-14849
Signed-off-by: Sivakumar Thulasimani <sivakumar.thulasimani@intel.com>
Reviewed-on: https://android.intel.com:443/468190
---
 drivers/gpu/drm/i915/i915_reg.h      | 7 +++++++
 drivers/gpu/drm/i915/intel_display.c | 9 +++++++++
 2 files changed, 16 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index 313192a..0d8e9e9 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -1094,6 +1094,13 @@ enum punit_power_well {
 #define  DPIO_CHV_INT_LOCK_THRESHOLD_SEL_COARSE	1 /* 1: coarse & 0 : fine  */
 #define CHV_PLL_DW9(ch) _PIPE(ch, _CHV_PLL_DW9_CH0, _CHV_PLL_DW9_CH1)
 
+#define _CHV_PLL_DW10_CH0		0x8028
+#define _CHV_PLL_DW10_CH1		0x81A8
+#define   DPIO_CHV_DCOAMP_OVERRIDE_EN	(1 << 27)
+#define   DPIO_CHV_DCOAMP_SHIFT		10
+#define   DPIO_CHV_DCOAMP_MASK		(0xF << DPIO_CHV_DCOAMP_SHIFT)
+#define CHV_PLL_DW10(ch) _PIPE(ch, _CHV_PLL_DW10_CH0, _CHV_PLL_DW10_CH1)
+
 #define _CHV_CMN_DW5_CH0               0x8114
 #define   CHV_BUFRIGHTENA1_DISABLE	(0 << 20)
 #define   CHV_BUFRIGHTENA1_NORMAL	(1 << 20)
diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 46f00a1..91c4cc5 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -6936,6 +6936,15 @@ static void chv_update_pll(struct intel_crtc *crtc)
 	dpio_val |= (tribuf_calcntr << DPIO_CHV_TDC_TARGET_CNT_SHIFT);
 	vlv_dpio_write(dev_priv, pipe, CHV_PLL_DW8(port), dpio_val);
 
+	dpio_val = vlv_dpio_read(dev_priv, pipe, CHV_PLL_DW10(port));
+	dpio_val &= ~(DPIO_CHV_DCOAMP_OVERRIDE_EN | DPIO_CHV_DCOAMP_MASK);
+
+	if (vco >= 5000000 && vco < 5400000) {
+		dpio_val |= (DPIO_CHV_DCOAMP_OVERRIDE_EN |
+			0xF << DPIO_CHV_DCOAMP_SHIFT);
+	}
+	vlv_dpio_write(dev_priv, pipe, CHV_PLL_DW10(port), dpio_val);
+
 	/* AFC Recal */
 	vlv_dpio_write(dev_priv, pipe, CHV_CMN_DW14(port),
 			vlv_dpio_read(dev_priv, pipe, CHV_CMN_DW14(port)) |
-- 
1.9.1

